src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -676,7 +676,7 @@ protected synchronized void initializeStorage() throws IOException {
                 h.setForDeserialize(storage.getNode(entry.getValue()));
                 heads.put(h.getId(), h);
             } else {
-                throw new IOException("Tried to load head FlowNodes for execution "+this.owner+" but FlowNode was not found in storage for head id:FlowNodeId "+entry.getKey()+":"+entry.getValue());
+                throw new IOException("Flow node could not be loaded because Jenkins did not shut down cleanly and the Pipeline Durability was not set to MAX_SURVIVABILITY");
```

Comment:
Also, this might happen even if you are using `MAX_SURVIVABILITY` in obscure cases, but I'm not sure. Maybe check the actual durability setting and only add the "and the Pipeline durability setting was not set to MAX_SURVIVABILITY" part of the message if that is in fact true. Probably only matters if anyone in the Jira issue mentioned seeing this exception when using `MAX_SURVIVABILITY`, otherwise it could be improved later as needed.

Reply:
I would also hesitate to suggest a remediation without actually knowing that it suffices.